packages/ngtools/webpack/src/ivy/plugin.ts

```diff
@@ -0,0 +1,486 @@
+          }
+
+          builder.getAllDependencies(sourceFile).forEach((dep) => currentUnused.delete(dep));
+        });
+        for (const unused of currentUnused) {
+          if (previousUnused && previousUnused.has(unused)) {
+            continue;
+          }
+          compilation.warnings.push(
+            `${unused} is part of the TypeScript compilation but it's unused.\n` +
+              `Add only entry points to the 'files' or 'include' properties in your tsconfig.`,
+          );
+        }
+        previousUnused = currentUnused;
+      });
+
+      // Store file emitter for loader usage
+      compilation[AngularPluginSymbol] = fileEmitter;
+    });
+  }
+
+  private updateAotProgram(
+    compilerOptions: CompilerOptions,
+    rootNames: string[],
+    host: CompilerHost,
+    diagnosticsReporter: DiagnosticsReporter,
+    resourceLoader: WebpackResourceLoader,
+  ) {
+    // Create an ngtsc plugin instance used to initialize a TypeScipt host and program
+    const ngtsc = new NgTscPlugin(compilerOptions);
+
+    const wrappedHost = ngtsc.wrapHost(host, rootNames, compilerOptions);
+
+    // SourceFile versions are required for builder programs.
+    // The wrapped host adds additional files that will not have versions
+    augmentHostWithVersioning(wrappedHost);
+
+    const oldProgram = this.ngtscNextProgram;
+    const program = ts.createProgram({
+      options: compilerOptions,
+      oldProgram,
+      rootNames: wrappedHost.inputFiles,
+      host: wrappedHost,
+    });
+
+    // The `ignoreForEmit` return value can be safely ignored when emitting. Only files
+    // that will be bundled (requested by Webpack) will be emitted. Combined with TypeScript's
+    // eliding of type only imports, this will cause type only files to be automatically ignored.
+    // Internal Angular type check files are also not resolvable by the bundler. Even if they
+    // were somehow errantly imported, the bundler would error before an emit was attempted.
+    // Diagnostics are still collected for all files which requires using `ignoreForDiagnostics`.
+    const { ignoreForDiagnostics, ignoreForEmit } = ngtsc.setupCompilation(program, oldProgram);
+
+    const builder = ts.createEmitAndSemanticDiagnosticsBuilderProgram(
+      program,
+      wrappedHost,
+      this.builder,
+    );
+
+    // Save for next rebuild
+    if (this.watchMode) {
+      this.ngtscNextProgram = ngtsc.getNextProgram();
+      this.builder = builder;
+    }
+
+    // Update semantic diagnostics cache
+    while (true) {
+      const result = builder.getSemanticDiagnosticsOfNextAffectedFile(undefined, (sourceFile) =>
+        ignoreForDiagnostics.has(sourceFile),
+      );
+      if (!result) {
+        break;
+      }
+    }
+
+    // Collect semantic diagnostics
+    for (const sourceFile of builder.getSourceFiles()) {
+      if (!ignoreForDiagnostics.has(sourceFile)) {
+        diagnosticsReporter(builder.getSemanticDiagnostics(sourceFile));
+      }
+    }
```

Comment:
I believe that the non-semantic diagnostics are typically gathered _before_ semantic diagnostics.